File: cli-node/src/repos.ts
/** Domain roots and repository discovery (real filesystem, no virtual tree). */

import { existsSync, readdirSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { basename, join, relative, resolve } from "node:path";

export type Kind = "domain" | "repo" | "dir";

export interface Candidate {
  path: string;
  domain: string;
  kind: Kind;
}

export function defaultRoots(): string[] {
  const override = process.env.ATLAS_ROOTS;
  if (override) return override.split(":").filter((p) => p.trim());
  return [join(homedir(), "@development"), join(homedir(), "@megavale-repos")];
}

function isGitRepo(path: string): boolean {
  return existsSync(join(path, ".git"));
}

function isDir(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

/** Split a path into [domain name, path relative to domain root]. Null domain when outside roots. */
export function splitDomain(path: string, roots: string[] = defaultRoots()): [string | null, string] {
  for (const root of roots) {
    const anchor = resolve(root);
    for (const candidate of [resolve(path)]) {
      if (candidate === anchor) return [basename(anchor), ""];
      if (candidate.startsWith(anchor + "/")) return [basename(anchor), relative(anchor, candidate)];
    }
  }
  return [null, path];
}

export function classify(path: string, roots: string[] = defaultRoots(), workspaces: Set<string> = new Set()): string {
  for (const root of roots) {
    if (resolve(path) === resolve(root)) return "domain";
  }
  if (workspaces.has(resolve(path))) return "workspace";
  if (isGitRepo(path)) return "repo";
  return "dir";
}

export function discover(roots: string[] = defaultRoots(), maxDepth = 2): Candidate[] {
  const found: Candidate[] = [];
  for (const root of roots) {
    if (!isDir(root)) continue;
    const domain = basename(root);
    found.push({ path: root, domain, kind: "domain" });
    let entries: string[];
    try {
      entries = readdirSync(root).sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
    } catch {
      continue;
    }
    for (const name of entries) {
      if (name.startsWith(".")) continue;
      const entry = join(root, name);
      if (!isDir(entry)) continue;
      const kind: Kind = isGitRepo(entry) ? "repo" : "dir";
      found.push({ path: entry, domain, kind });
      if (maxDepth >= 2 && kind === "dir") {
        let subs: string[];
        try {
          subs = readdirSync(entry).sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
        } catch {
          continue;
        }
        for (const sub of subs) {
          if (sub.startsWith(".")) continue;
          const subPath = join(entry, sub);
          if (isDir(subPath) && isGitRepo(subPath)) found.push({ path: subPath, domain, kind: "repo" });
        }
      }
    }
  }
  return found;
}

File: cli/atlas_cli/repos.py
"""Domain roots and repository discovery (real filesystem, no virtual tree)."""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path


@dataclass
class Candidate:
    path: str
    domain: str  # display name, e.g. "@megavale-repos"
    kind: str  # "domain" | "repo" | "dir"


def default_roots() -> list[Path]:
    override = os.environ.get("ATLAS_ROOTS")
    if override:
        return [Path(p).expanduser() for p in override.split(":") if p.strip()]
    home = Path.home()
    return [home / "@development", home / "@megavale-repos"]


def split_domain(path: str, roots: list[Path] | None = None) -> tuple[str | None, str]:
    """Split a path into (domain name, path relative to the domain root).

    Returns (None, path) when the path is outside every known root.
    The relative part is "" when path is the domain root itself.
    """
    roots = roots if roots is not None else default_roots()
    target = Path(path)
    candidates = [target]
    if target.exists():
        candidates.append(target.resolve())
    for root in roots:
        anchor = root.resolve() if root.exists() else root
        for candidate in candidates:
            try:
                rel = candidate.relative_to(anchor)
            except ValueError:
                continue
            return (root.name, "" if str(rel) == "." else str(rel))
    return (None, path)


def classify(path: Path, roots: list[Path], workspaces: set[str] | None = None) -> str:
    """Classify a real path as domain, workspace, repo, or plain dir.

    Mirrors the brief: a workspace is recognised only when the path matches a
    registered bookmark, never by folder name alone.
    """
    resolved = str(path.resolve()) if path.exists() else str(path)
    for root in roots:
        try:
            if path.resolve() == root.resolve():
                return "domain"
        except OSError:
            continue
    if workspaces and resolved in workspaces:
        return "workspace"
    if (path / ".git").exists():
        return "repo"
    return "dir"


def discover(roots: list[Path] | None = None, max_depth: int = 2) -> list[Candidate]:
    """List domain roots plus repos/dirs up to max_depth (depth 1 always listed)."""
    roots = roots if roots is not None else default_roots()
    found: list[Candidate] = []
    for root in roots:
        if not root.is_dir():
            continue
        domain = root.name
        found.append(Candidate(path=str(root), domain=domain, kind="domain"))
        try:
            entries = sorted(root.iterdir(), key=lambda p: p.name.lower())
        except OSError:
            continue
        for entry in entries:
            if entry.name.startswith(".") or not entry.is_dir():
                continue
            kind = "repo" if (entry / ".git").is_dir() or (entry / ".git").is_file() else "dir"
            found.append(Candidate(path=str(entry), domain=domain, kind=kind))
            if max_depth >= 2 and kind == "dir":
                try:
                    subentries = sorted(entry.iterdir(), key=lambda p: p.name.lower())
                except OSError:
                    continue
                for sub in subentries:
                    if sub.name.startswith(".") or not sub.is_dir():
                        continue
                    if (sub / ".git").is_dir() or (sub / ".git").is_file():
                        found.append(Candidate(path=str(sub), domain=domain, kind="repo"))
    return found
